### ISSUE TITLE
Smaller cosign-verify binary

### ISSUE THREAD

**Issue by steiza** [sigstore/cosign#3961]
**Description**

The compiled cosign binary is quite large (~100 MB) and unlikely to get dramatically smaller in the near term. Part of this is due to the flexibility of cosign, in the number of different ways you can sign content, which our users appreciate.

But at the same time, ecosystems adopting Sigstore might want a smaller, self-contained, utility specifically for verifying Sigstore content. This repository could produce a smaller binary, named something like `cosign-verify`, that just handles the verification commands `cosign` supports, and hopefully would be an order of magnitude smaller (~10 MB).

**Comment by codysoyland** [comment]
Is the primary goal to make a smaller binary? I wonder if the work being done in https://github.com/sigstore/sigstore/pull/1901 for making KMS providers pluggable could reduce it substantially, or we there are other means to reducing cosign's binary size.

Would it try to mirror the verification commands/flags that exist in cosign today, or would verification be rewritten to solely use sigstore-go and simplify the flags?

If this is a simplified rewrite, I worry how much this overlaps/conflicts with the idea of migrating the cosign commands to support trustedroots/bundles natively and deprecating existing functionality. This proposal could be the start of a "rewrite" of cosign, and I worry a bit about fragmentation when maintaining two command sets with overlapping functionality.

I'm definitely in favor of having a cosign verifier that is much smaller, and perhaps a rewrite is the right answer, but I want to understand how this fits into cosign's strategy and if we could accomplish the same goals while improving the overall cosign codebase.

**Comment by steiza** [comment]
Yes, so if we did this we would definitely want to do it in a way that was minimally disruptive to the cosign codebase and this would definitely **not** be a rewrite.

I haven't been thinking about this for very long, but for example, what if we split out [the way we add subcommands to `cosign`](https://github.com/sigstore/cosign/blob/8f0a179c88b864eba58ee425b9ae064c381a925c/cmd/cosign/cli/commands.go#L94) into verifier and non-verifier parts? `cosign` would still do both, but then we'd have a codepath for a `cosign-verify` command that has the same API, but reduced functionality and (hopefully if the go compiler's dead code detection is good) a much smaller footprint on disk.

I'm not committing to that exact implementation strategy (I want to think about it more) but that's an example of how we could achieve this without a lot of work upfront or maintenance work in the future.

**Comment by codysoyland** [comment]
It would definitely be interesting to see how far we can get with only a minor refactor and deadcode detection. If it used the same codebase, that would be a huge win for maintainability.

**Comment by ChristianCiach** [comment]
Related issue:

*  https://github.com/sigstore/cosign/issues/1462

**Comment by haydentherapper** [comment]
I'm supportive of this! A few thoughts:

* This verifier should be written based on sigstore-go and we should further prune dependencies there
* I think this verifier should be minimal in scope - support only artifacts and attestations.
  * For containers or other metadata that requires canonicalization (and large dependencies to calculate those hashes), we could support passing the hash for verification.
  * For verifying with KMS, we'll rely on KMS plugins.
* I'd like to avoid duplication between sigstore-go's example binary and Cosign. I think we could move the example verifier out from sigstore-go into Cosign, and use the same CLI flag language in Cosign for this new binary.
* For further duplication reduction, we likely can heavily rely on the new-trust-bundle verification codepath in Cosign

**Comment by ramonpetgrave64** [comment]
@codysoyland re: https://github.com/sigstore/sigstore/pull/1901, I tested a bit with the dependencies, removing the existing `signature/kms/[aws/azure/gcp/valut]` imports from sigstore, cosign, and timestamp-authority. Rebuilding the macos binary went from 105MB to 97MB.

I tried using `https://github.com/jondot/goweight`, and it seems the biggest contributors are

```
   23 MB k8s.io/api/core/v1
   17 MB github.com/google/go-github/v55/github
   13 MB runtime
   12 MB github.com/xanzy/go-gitlab
   11 MB github.com/alibabacloud-go/cr-20181201/client
   11 MB github.com/aws/aws-sdk-go-v2/service/kms
  9.8 MB github.com/open-policy-agent/opa/ast
  9.6 MB github.com/aws/aws-sdk-go-v2/service/ecr
  8.7 MB net/http
  7.6 MB github.com/open-policy-agent/opa/topdown
...
```

**Comment by codysoyland** [comment]
@ramonpetgrave64 This is great data, thanks

**Comment by Hayden-IO** [comment]
We've found that we can't significantly reduce verify much smaller than sign, so we're going to move forward with trying to reduce the existing Cosign binary over time.
